### ISSUE TITLE
PS-PurrSoftware

### ISSUE THREAD

**Issue by knchaffin** [VCVRack/community#624]
https://github.com/knchaffin/Meander

**Comment by cschol** [comment]
Before I integrate this into the library, one question: the repository is called Meander. Are you expecting to have more than one plugin? Then I would rename the repository to something resembling your brand (or slug).

**Comment by knchaffin** [comment]
I do not plan to develop more plugins.  Also, the architecture of Meander is such that no other modules could share the same plugin with Meander, just as Meander can only have one fully functional instance of the module running.  Meander can have a 2nd or more instance loaded but the code is set up with Andrew’s recommended soliton global variable to limit to one functional instance and other instances with a reduced panel size, no functional module widget and warning text on the panel.  This is due to Meander being an existing application with a lot of global data that is not thread safe.

So, let’s keep the repository named Meander and in the unlikely event that I develop other modules, they would be in a separate plugin and repository.  

Thanks, 
Ken Chaffin

**Comment by AndrewBelt** [comment]
>no other modules could share the same plugin with Meander

I'm not sure why that would be true.
Regardless, the repo name doesn't matter a whole lot. It can be changed at any time, unlike plugin/module slugs.

**Comment by knchaffin** [comment]
Yeah, I'm still a unclear on sharing global data and code between two modules in the same plugin.  The main thing is that two modules in the same plugin should not both include meander.hpp, I think, as that has non thread safe functions in it.  I suppose there could be another global header at the plugin/DLL level for global data that is safe to share.

I'm still learning.  Thanks for clarifying.

**Comment by knchaffin** [comment]
Yep, I'm still learning.  I accidently closed the issue when I responded. Sorry about that.

**Comment by cschol** [comment]
`DEBUG` statements in code: https://github.com/knchaffin/Meander/issues/5

**Comment by knchaffin** [comment]
In file meander.hpp lines 3 and 4 are:

`// the following will "comment out" all DEBUG() calls for speed.  Comment out the next line and DEBUG is compiled
#define DEBUG(format, ...) // DEBUG(format, ...)

This line redefines DEBUG such that DEBUG statements are not compiled,  How else should I "disable via conditional compilation"?

**Comment by knchaffin** [comment]
I've looked at other's source code and see how others do either conditional compilation or conditional function call for DEBUG() .  My method accomplishes the same thing.

**Comment by knchaffin** [comment]
Please let me know what the acceptable ways to handle this are.  Is my current DEBUG() redefinition acceptable as it results in the DEBUG() calls being seen by the compiler as comment lines and thus are not compiled?  If not acceptable, are conditional calls acceptable, such that the compiler compiles the DEBUG calls but they are never executed?  Some plugins in the library do use this method.  Or, if the only acceptable methods are to put conditional compile directives around each DEBUG call, let me know.

Thanks

**Comment by cschol** [comment]
Your method is fine (and is what I meant by conditional compilation). But, it does not seem to work. I still see `DEBUG` messages in the log when I compile and run your plugin as it is in the repository. Please look into that.

**Comment by knchaffin** [comment]
Okay, I have now changed to the technique used by other modules in the library of the type:
if (doDebug) DEBUG();
doDebug is a global bool set to false in the repository code.
No debug statements are showing up in the log.
Changes are committed to GitHub.

**Comment by cschol** [comment]
That works. Thank you!

**Comment by knchaffin** [comment]
Please update PS-PurrSoftware Meander to V1.0.2 per
commit 8cff0277ba218e9e4c51b607f53a847274b0da39

This version primarily adds the manual and source file URLs to my plugin.json file.  A fairly lengthy manual was created.

Thanks.

**Comment by knchaffin** [comment]
If possible, please use commit 70d56ad855254bc1f6ccb675d59aa4aed8aa9d49 
which has a slight bug fix.  If too late, that's okay.

**Comment by AndrewBelt** [comment]
Built 8cff0277ba218e9e4c51b607f53a847274b0da39, but if you'd like to release 70d56ad855254bc1f6ccb675d59aa4aed8aa9d49, bump the version and request an update.

**Comment by knchaffin** [comment]
Please release plugin V1.0.3 to commit 1019b32ce7a7a13be3f1e3a5ca5a02d4fe690f3a

**Comment by knchaffin** [comment]
Please update PurrSoftware Meander to V1.0.4 per commit 8420c666d4cc982bc7a58dfdba93cdbf40b4eb57

Thanks

**Comment by knchaffin** [comment]
Please update PurrSoftware Meander to V1.0.4 per commit ee06cefbdaf82fa33f87936232ac90b2105dab49

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander to
commit 19d7265ca46703d98e89c6bd3bbf70b5aa115035

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander to V1.0.6 per 
commit b5aebe177de9a052bdedabe4d16b0debc12a6b9f

This is a bug fix release.

Thanks

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander to V1.0.7 per
commit ed0119856672ed60d2b93d0d7fcab929dc9c35f1

These should be the last fixes for plugin browser and Library appearance bugs.

Thanks

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander to V1.0.8 per
commit 490526193f74fa7cb2852f85315aee38ca389045

Thanks

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander to V1.0.9 per
commit 867424607d05c238356015c02f9ffb47b8254aa0

Thanks

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander to V1.0.10 per
commit d205470b7990b6dd0d04aba4d23632938c4b152b

Thanks

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander to V1.0.11 per
commit f14084d2fe568f7448f49a4ef4345d9a53d42df5

Thanks

**Comment by knchaffin** [comment]
Please add PurrSoftware Meander to Rack V2 library.
https://github.com/knchaffin/Meander/blob/master-V2/
dc9001e commit b83befc62d7f475dbdeaee87b859ef1d99ab1ebf

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander V2.0.13 for the Rack V2  library.
https://github.com/knchaffin/Meander/blob/master-V2/ 
bc290de commit 3dafdbda0a731b514c226475dec47bbf32698ad0

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander V2.0.14 for the Rack V2  library.
https://github.com/knchaffin/Meander/blob/master-V2/ 
[23e7be0](https://github.com/knchaffin/Meander/commit/23e7be0d1703c2d9ca78c8b57d5a6e12b498d787) commit 4ae64a27f5021a2436661892febf872df4bf3f8b

**Comment by knchaffin** [comment]
Please build PurrSoftware Meander V2.0.15 for the Rack V2 library.
[5fc4ea4](https://github.com/knchaffin/Meander/commit/5fc4ea46e882852597afd738edb48cab1a0a2790) commit dae2470d89833a654a816434a4e3d65f6fdfaec4

**Comment by AndrewBelt** [comment]
The panel of Meander appears to not load on Linux. Is the filename correct?

**Comment by knchaffin** [comment]
The filenames seem correct, including case.  When you say the panel appears to not load, is that in the browser or in the patch?  If there is a module, Meander loads the SVG to get panel dimensions but does procedural panel rendering.  If the module is null, the SVG is still loaded for the panel size but a light are dark theme PNG file is loaded in MeanderWidget::draw() override via:
std::shared_ptr<Image> lightPanelImage=lightPanelImage = APP->window->loadImage(asset::plugin(pluginInstance,"res/Meander-light.png"));  // for example

Ah, I  have   lightPanelImage=lightPanelImage = 

I will do a commit of this correction, but, I want to commit a couple of changes I have made during my testing after I committed this build commit.  Is that okay or do we need to close this current issue and then I update and commit and request the issue be reopened?

**Comment by knchaffin** [comment]
Is this a problem:  APP->window->loadImage(() ?
I.E., should I leave APP->window off as I see one developer does when loading PNGs?

**Comment by knchaffin** [comment]
I have committed my fixes via 1 parent [dae2470](https://github.com/knchaffin/Meander/commit/dae2470d89833a654a816434a4e3d65f6fdfaec4) commit e35fed7a167ef59e58bc0e0025963a5a14378f1c

**Comment by knchaffin** [comment]
I have my doubts that this commit will fix the problem Andrew points out.  Whereas I had a typo where I did a multiple assignment of an Image pointer type to itself, that should not cause problems unless Linux handles this differently than Windows where I develop.  If the panel still "appears to not load on Linux", I will need some feedback as to exactly what that means so that I know where to look for a problem.  Thanks.

**Comment by cschol** [comment]
Panel loads correctly on Linux in module browser and in Rack.

**Comment by knchaffin** [comment]
Thanks for the feedback!  I can relax a bit now, I hope.

**Comment by knchaffin** [comment]
Please build Purr Software Meander V2.0.16 cosmetic fix per:
https://github.com/knchaffin/Meander/commit/8092eb8f0ee51b40c43fdc1e1e54c1c09f622a9b

**Comment by knchaffin** [comment]
Please build Purr Software Meander enhancement release V2.0.17 for the library, per:
1 parent [2e77fc4](https://github.com/knchaffin/Meander/commit/2e77fc4eefeeef4ac7bea9c51e54564f86b3b1d8) commit 8a81a4df886c4c2cd7de3aaf721fbf848bba6f93

**Comment by knchaffin** [comment]
Please build PS Software Meander for the library per:
1 parent [784287e](https://github.com/knchaffin/Meander/commit/784287ec2e0c37f5dbc9abb4070241af086198a1) commit 1dd43428ae1f5c774588c5fda224e494de631773

Removes a possible font license issue.

**Comment by knchaffin** [comment]
Please build PS Software Meander for the library per:
1 parent [a60e882](https://github.com/knchaffin/Meander/commit/a60e88261e3bd8ad43417f9a08f8317c1cd4c47e) commit 30976ac966094bb700771bc5fb0742fcd770d24d

Removes a possible font license issue and fix pentatonic scale output.

**Comment by knchaffin** [comment]
Please build PS PurrsSoftware Meander for the library per:
1 parent [af1841d](https://github.com/knchaffin/Meander/commit/af1841d1d64fac219fd024d9169d147d9073ea5d) commit c095824708947630d9db6a6b7afcd51bdaa0a009

This update has font changes only.

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.20 enhancement release for the library per:
1 parent [32f2574](https://github.com/knchaffin/Meander/commit/32f25742ce676ca7d05d699e495a12351c0955b1) commit 36ce1464405b04f2b5cf9e5c5becaa9a75fbe1b7

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.21 enhancement release for the library per:
1 parent [b65d23d](https://github.com/knchaffin/Meander/commit/b65d23d7396b66be8ade514e13d63db67dd5f1ca) commit 2ccf06b6932a0baf57d424cc9208faccfc7e7a57

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.22 enhancement release for the library per:
1 parent [d03d200](https://github.com/knchaffin/Meander/commit/d03d2009a2980522a6f32818f362c0c132233b70) commit dabf0e1991e991599e4482c71f4c1df8c663fdf8

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.23 enhancement release for the library per:
1 parent [5ebaafa](https://github.com/knchaffin/Meander/commit/5ebaafa58cf5e4232a367d6761d190047f2796b7) commit 16a802fb6d6bcf872672a2a8b57aa07f18630069

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.24 enhancement release for the library per:
1 parent [23612c5](https://github.com/knchaffin/Meander/commit/23612c59ae5173738ca81474b0226594db2dcd76) commit a7d12b914c57b55bce1597503ef810fbf8d5afc0

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.25 enhancement release for the library per:
1 parent [37b2f47](https://github.com/knchaffin/Meander/commit/37b2f474ede1c77aa1f16df02415dc8134842626) commit f346750a889a429c2b5ee7bf74c0983dd5450022

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.26 minor bug-fix release for the library per:
1 parent [d30882d](https://github.com/knchaffin/Meander/commit/d30882d907754b2af6aeaa37f8055e1df8e64095) commit 20e4681d3bc141d33146189089ed13e90c15aa16

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.27 minor enhancement release for the library per:
1 parent [abb4cfc](https://github.com/knchaffin/Meander/commit/abb4cfc3137688d2b5ac58feb53c3c11ed7d627b) commit f4d9145176a6121f5ac377cb7fe3498678878efe

**Comment by knchaffin** [comment]
Please build PS PurrSoftware Meander V.2.0.28 minor enhancement release for the library per:
1 parent [7cc89bd](https://github.com/knchaffin/Meander/commit/7cc89bd1cf9ca42e1790a1deb06b2099980668fa) commit 3ad2248

**Comment by knchaffin** [comment]
Please build PS PurrSoftware  V.2.0.29  plugin minor enhancement for Meander and new module ModeSeqQuant release for the library per:
1 parent [f90da6e](https://github.com/knchaffin/Meander/commit/f90da6e70e7128cf4130adcf3f683d8daaca41cb) commit bb4d39f

I am aware that my GitHub repository still refers to Meander as the root, but this works fine for the PS PurrSoftware plugin.

**Comment by knchaffin** [comment]
Please build PS PurrSoftware  V.2.0.30  plugin minor enhancement and cosmetic user experience improvement release for the library per:
1 parent [f219b15](https://github.com/knchaffin/Meander/commit/f219b15a39523b326105b8c30ad456501e5e0e80) commit 2ce7570

I am aware that my GitHub repository still refers to Meander as the root, but this works fine for the PS PurrSoftware plugin.

**Comment by knchaffin** [comment]
Please build PS PurrSoftware V.2.0.31 plugin minor cosmetic user experience fix release for the library per:

1 parent [f3c9d36](https://github.com/knchaffin/Meander/commit/f3c9d363a1a6e7a6ecdd7d603029518a4996871f) commit ac4f5da

**Comment by knchaffin** [comment]
Please build PS PurrSoftware V.2.0.32 plugin with one new module ModeScaleProgressions and UI/UE fixes for 3 modules.

1 parent [f927ae5](https://github.com/knchaffin/Meander/commit/f927ae5bf300281e3f4cf1f97cd57bf52da0dad7) commit fa5148b

**Comment by knchaffin** [comment]
Please build PS PurrSoftware V.2.0.33 plugin with cosmetic fixes only..

1 parent [10d5afb](https://github.com/knchaffin/Meander/commit/10d5afbedef8665eb061024eee7ecbd2bff6a84a) commit 55a9f33

**Comment by knchaffin** [comment]
Please build PS PurrSoftware V.2.0.34 plugin with critical bug fix that was affecting the browser experience.

1 parent [9b0a522](https://github.com/knchaffin/Meander/commit/9b0a5223ac00ee137975d656b0a699d652c37024) commit 71968f4

**Comment by knchaffin** [comment]
Please build PS PurrSoftware V.2.0.35 plugin with minor user requested fixes.

1 parent 
[e0a8cf7](https://github.com/knchaffin/Meander/commit/e0a8cf7c3f1caf8c778b6935e9cde903c0e6502b)
 commit b4cec99
